### PR TITLE
Refactor code to incorporate object tree related changes

### PR DESCRIPTION
**Pull request by cybernth**
tendrl-bug-id: Tendrl/ceph-integration/issues#85
Signed-off-by: Nishanth Thomas <nthomas@redhat.com>

### CODE REVIEW

File: tendrl/ceph_integration/__init__.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+__import__('pkg_resources').declare_namespace(__name__)
```
Comment (r0h4n) [inline tendrl/ceph_integration/__init__.py]: remove this 1st statement

File: etc/tendrl/ceph-integration/ceph-integration.conf.yaml.sample [hunk 2]
```diff
@@ -0,0 +1,5 @@
+etcd_port: 2379
+etcd_connection: 0.0.0.0
+tendrl_exe_file_prefix: $HOME/.tendrl/util_runner_
```
Comment (r0h4n) [inline etc/tendrl/ceph-integration/ceph-integration.conf.yaml.sample]: https://github.com/Tendrl/node-agent/blob/master/etc/tendrl/node-agent/node-agent.conf.yaml.sample#L4

File: tendrl/__init__.py [hunk 3]
```diff
@@ -1 +0,0 @@
-__import__('pkg_resources').declare_namespace(__name__)
```
Comment (r0h4n) [inline tendrl/__init__.py:1]: add this back, this will break the entire python namespace

File: tendrl/ceph_integration/__init__.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+try:
```
Comment (r0h4n) [inline tendrl/ceph_integration/__init__.py:2]: https://github.com/Tendrl/node-agent/blob/master/tendrl/node_agent/__init__.py#L1

File: tendrl/ceph_integration/central_store/__init__.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+from tendrl.commons.central_store import EtcdCentralStore
+
+
+class CephIntegrationEtcdPersister(EtcdCentralStore):
+    def __init__(self):
+        super(CephIntegrationEtcdPersister, self).__init__()
+
```
Comment (r0h4n) [inline tendrl/ceph_integration/central_store/__init__.py:7]: ensure "save_%lower(objname)"  is the function name

File: tendrl/ceph_integration/flows/__init__.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import logging
+
+from tendrl.commons import flows
+
+LOG = logging.getLogger(__name__)
```
Comment (r0h4n) [inline tendrl/ceph_integration/flows/__init__.py]: why do you need a LOG?

File: tendrl/ceph_integration/flows/create_pool.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
-from tendrl.commons.flows.base_flow import BaseFlow
+from tendrl.ceph_integration.flows import CephIntegrationBaseFlow
```
Comment (r0h4n) [inline tendrl/ceph_integration/flows/create_pool.py]: This file should be tendrl/ceph_integration/flows/create_pool/ __init__.py , and put this class in it, class name should match flow name as in definition file

File: tendrl/ceph_integration/manager/__init__.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+from tendrl.ceph_integration.manager import utils
+from tendrl.ceph_integration.central_store import CephIntegrationEtcdPersister
+from tendrl.commons.manager import Manager
+from tendrl.commons.sds_sync import SdsSyncThread
+
+
+LOG = logging.getLogger(__name__)
+
+
+class CephIntegrationSyncStateThread(SdsSyncThread):
```
Comment (r0h4n) [inline tendrl/ceph_integration/manager/__init__.py]: - change to CephIntegrationSdsSyncStateThread
- put this class in a tendrl/ceph_integration/sds_sync/__init__.py   (check https://github.com/Tendrl/node-agent/blob/master/tendrl/node_agent/node_sync/__init__.py)

File: tendrl/ceph_integration/manager/__init__.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+    def __init__(self, manager):
+        super(CephIntegrationSyncStateThread, self).__init__()
+
+        self._manager = manager
```
Comment (r0h4n) [inline tendrl/ceph_integration/manager/__init__.py]: remove manager and complete

https://github.com/Tendrl/node-agent/blob/master/tendrl/node_agent/node_sync/__init__.py#L13

File: tendrl/ceph_integration/manager/__init__.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+        while not self._complete.is_set():
+            data = ceph.heartbeat(heartbeat_type="cluster", discover=True)
+            if data is not None:
+                for tag, cluster_data in data.iteritems():
+                    try:
+                        if tag.startswith("ceph/cluster/"):
+                            if not cluster_data[
+                                    'fsid'
+                            ] in self._manager.clusters:
```
Comment (r0h4n) [inline tendrl/ceph_integration/manager/__init__.py]: this self._manager.on_pull has to go completely, refactor to match node-agent's on pull

https://github.com/Tendrl/node-agent/blob/master/tendrl/node_agent/node_sync/__init__.py#L13

File: tendrl/ceph_integration/manager/__init__.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+
+class CephIntegrationManager(Manager):
+    def __init__(self):
+        self._complete = gevent.event.Event()
+        super(
+            CephIntegrationManager,
+            self
+        ).__init__(
+            CephIntegrationSyncStateThread(self),
```
Comment (r0h4n) [inline tendrl/ceph_integration/manager/__init__.py]: dont pass "self"

File: tendrl/ceph_integration/manager/__init__.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        del self.clusters[fs_id]
+
+        self._expunge(fs_id)
+
+    def join(self):
+        super(self).join()
+        for monitor in self.clusters.values():
+            monitor.join()
+
+    def on_pull(self, heartbeat_data, cluster_id):
```
Comment (r0h4n) [inline tendrl/ceph_integration/manager/__init__.py]: move this to your  sdssyncthread

File: tendrl/ceph_integration/manager/__init__.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        # by the monitor.
+        cluster_monitor.start()
+        # Wait for ClusterMonitor to start accepting events before asking it
+        # to do anything
+        cluster_monitor.ready()
+        cluster_monitor.on_heartbeat(heartbeat_data['id'], heartbeat_data)
+
+    def register_to_cluster(self):
+        # save tendrl context
+        tendrl_ns.tendrl_context.save(tendrl_ns.etcd_orm)
```
Comment (r0h4n) [inline tendrl/ceph_integration/manager/__init__.py]: remove the argument

File: tendrl/ceph_integration/manager/__init__.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+            gevent.sleep(3)
+
+        LOG.info("%s complete" % self.__class__.__name__)
+
+
+class CephIntegrationManager(Manager):
+    def __init__(self):
+        self._complete = gevent.event.Event()
+        super(
+            CephIntegrationManager,
```
Comment (r0h4n) [inline tendrl/ceph_integration/manager/__init__.py:17]: https://github.com/Tendrl/node-agent/blob/master/tendrl/node_agent/manager/__init__.py#L20
https://github.com/Tendrl/node-agent/blob/master/tendrl/node_agent/manager/__init__.py#L79

File: tendrl/ceph_integration/objects/__init__.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+        # 'help': 'Checks if a node is up', 'outputs': ['Node.status'],
+        # 'enabled': True, 'name': 'check whether the node is up',
+        # 'type': 'Create', 'uuid': 'eda0b13a-7362-48d5-b5ca-4b6d6533a5ab'}}
+        self.atoms = atoms or obj_def.atoms
+
+        # List of flows under this object
+        self.flows = flows or obj_def.flows
+
+
+class NodeAgentBaseAtom(atoms.BaseAtom):
```
Comment (r0h4n) [inline tendrl/ceph_integration/objects/__init__.py]: rename

File: tendrl/ceph_integration/objects/definition/__init__.py [hunk 16]
```diff
@@ -0,0 +1,69 @@
+import importlib
```
Comment (r0h4n) [inline tendrl/ceph_integration/objects/definition/__init__.py:1]: recheck with https://github.com/Tendrl/node-agent/blob/master/tendrl/node_agent/objects/definition/__init__.py

File: tendrl/ceph_integration/objects/event/__init__.py [hunk 17]
```diff
@@ -0,0 +1,37 @@
+import logging
+
+from tendrl.commons.etcdobj.etcdobj import EtcdObj
+from tendrl.ceph_integration import objects
+
+
+LOG = logging.getLogger(__name__)
```
Comment (r0h4n) [inline tendrl/ceph_integration/objects/event/__init__.py]: remove LOG wherever unused

File: tendrl/ceph_integration/objects/event/__init__.py [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+class Event(objects.CephIntegrationBaseObject):
+    def __init__(self, id=None, when=None,
+                 severity=None, message=None, fsid=None,
+                 fqdn=None, service_type=None, service_id=None,
+                 *args, **kwargs):
+        super(Event, self).__init__(*args, **kwargs)
+
+        self.value = 'clusters/%s/events/%s'
+        self.id = id
```
Comment (r0h4n) [inline tendrl/ceph_integration/objects/event/__init__.py]: rename to event_id, id causes namespace collisions

File: tendrl/ceph_integration/objects/pool/__init__.py [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+
+
+class Pool(objects.CephIntegrationBaseObject):
+    def __init__(self, cluster_id=None, pool_id=None,
+                 poolname=None, pg_num=None, min_size=None,
+                 used=None, percent_used=None, *args, **kwargs):
+        super(Pool, self).__init__(*args, **kwargs)
+
+        self.value = 'clusters/%s/Pools/%s'
+        self.cluster_id = cluster_id
```
Comment (r0h4n) [inline tendrl/ceph_integration/objects/pool/__init__.py]: store/get cluster_id and fsid from tendrl_ns.tendrl_context.cluster_id/fsid wherever required. remove self.cluster_id

File: tendrl/ceph_integration/objects/pool/__init__.py [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+
+
+class _Pool(EtcdObj):
+    """A table of the Pool, lazily updated
+    """
+    __name__ = 'clusters/%s/Pools/%s'
+    _tendrl_cls = Pool
+
+    def render(self):
+        self.__name__ = self.__name__ % (self.cluster_id, self.pool_id)
```
Comment (r0h4n) [inline tendrl/ceph_integration/objects/pool/__init__.py]: same here, use tendrl_ns for cluster_id

File: tendrl/ceph_integration/objects/pool/atoms/create.py [hunk 21]
```diff
@@ -1,9 +1,9 @@
 from tendrl.ceph_integration.manager.crud import Crud
 from tendrl.ceph_integration.manager import utils as manager_utils
-from tendrl.commons.atoms.base_atom import BaseAtom
+from tendrl.ceph_integration import objects
 
 
-class Create(BaseAtom):
+class Create(objects.NodeAgentBaseAtom):
```
Comment (r0h4n) [inline tendrl/ceph_integration/objects/pool/atoms/create.py]: NodeAgentBaseAtom?

File: tendrl/ceph_integration/objects/pool/atoms/create.py [hunk 22]
```diff
@@ -1,9 +1,9 @@
...
     def run(self, parameters):
```
Comment (r0h4n) [inline tendrl/ceph_integration/objects/pool/atoms/create.py:11]: where is the "obj = Pool" statement?

File: tendrl/ceph_integration/objects/pool/atoms/delete.py [hunk 23]
```diff
@@ -1,9 +1,9 @@
 from tendrl.ceph_integration.manager.crud import Crud
 from tendrl.ceph_integration.manager import utils as manager_utils
-from tendrl.commons.atoms.base_atom import BaseAtom
+from tendrl.ceph_integration import objects
```
Comment (r0h4n) [inline tendrl/ceph_integration/objects/pool/atoms/delete.py:3]: fix this atom as per create atom comments

File: tendrl/ceph_integration/objects/pool/flows/delete_pool.py [hunk 24]
```diff
@@ -1,6 +1,6 @@
-from tendrl.commons.flows.base_flow import BaseFlow
+from tendrl.ceph_integration.flows import CephIntegrationBaseFlow
```
Comment (r0h4n) [inline tendrl/ceph_integration/objects/pool/flows/delete_pool.py]: fix this flow too as per create atom comments

File: tendrl/ceph_integration/objects/sync_object/__init__.py [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+
+
+class _SyncObject(EtcdObj):
+    """A table of the _Service, lazily updated
+    """
+    __name__ = 'clusters/%s/maps/%s'
+    _tendrl_cls = SyncObject
+
+    def render(self):
+        self.__name__ = self.__name__ % (self.cluster_id, self.sync_type)
```
Comment (r0h4n) [inline tendrl/ceph_integration/objects/sync_object/__init__.py]: store/get cluster_id from tendrl_ns

File: tendrl/ceph_integration/objects/tendrl_context/__init__.py [hunk 26]
```diff
@@ -0,0 +1,66 @@
+import logging
+import os
+
```
Comment (r0h4n) [inline tendrl/ceph_integration/objects/tendrl_context/__init__.py:3]: recheck https://github.com/Tendrl/node-agent/blob/master/tendrl/node_agent/objects/tendrl_context/__init__.py

File: tendrl/ceph_integration/objects/tendrl_context/__init__.py [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+    def __init__(self, integration_id=None, node_id=None,
+                 sds_name=None, sds_name=None, *args, **kwargs):
+        super(TendrlContext, self).__init__(*args, **kwargs)
+
+        self.value = 'clusters/%s/TendrlContext'
+
+        # integration_id is the Tendrl generated cluster UUID
+        self.cluster_id = integration_id or \
+            self._get_local_integration_id(node_id)
+        self.sds_name = sds_name or self._get_sds_name()
```
Comment (r0h4n) [inline tendrl/ceph_integration/objects/tendrl_context/__init__.py:22]: we have stopped putting sds details in tendrl_context

**Review by r0h4n — CHANGES_REQUESTED** [review]

File: etc/tendrl/ceph-integration/ceph-integration.conf.yaml.sample (hunk 2 above)
Comment (cybernth) [reply etc/tendrl/ceph-integration/ceph-integration.conf.yaml.sample]: Done

File: tendrl/__init__.py (hunk 3 above)
Comment (cybernth) [reply tendrl/__init__.py:1]: done

File: tendrl/ceph_integration/central_store/__init__.py (hunk 5 above)
Comment (cybernth) [reply tendrl/ceph_integration/central_store/__init__.py:7]: done

File: tendrl/ceph_integration/flows/__init__.py (hunk 6 above)
Comment (cybernth) [reply tendrl/ceph_integration/flows/__init__.py]: removed

File: tendrl/ceph_integration/flows/create_pool.py (hunk 7 above)
Comment (cybernth) [reply tendrl/ceph_integration/flows/create_pool.py]: done

File: tendrl/ceph_integration/manager/__init__.py (hunk 8 above)
Comment (cybernth) [reply tendrl/ceph_integration/manager/__init__.py]: done

File: tendrl/ceph_integration/manager/__init__.py (hunk 9 above)
Comment (cybernth) [reply tendrl/ceph_integration/manager/__init__.py]: done

File: tendrl/ceph_integration/manager/__init__.py (hunk 14 above)
Comment (cybernth) [reply tendrl/ceph_integration/manager/__init__.py:17]: done

File: tendrl/ceph_integration/manager/__init__.py (hunk 11 above)
Comment (cybernth) [reply tendrl/ceph_integration/manager/__init__.py]: This part is changed

File: tendrl/ceph_integration/manager/__init__.py (hunk 12 above)
Comment (cybernth) [reply tendrl/ceph_integration/manager/__init__.py]: done

File: tendrl/ceph_integration/manager/__init__.py (hunk 13 above)
Comment (cybernth) [reply tendrl/ceph_integration/manager/__init__.py]: done

File: tendrl/ceph_integration/objects/definition/__init__.py (hunk 16 above)
Comment (cybernth) [reply tendrl/ceph_integration/objects/definition/__init__.py:1]: done

File: tendrl/ceph_integration/objects/event/__init__.py (hunk 17 above)
Comment (cybernth) [reply tendrl/ceph_integration/objects/event/__init__.py]: done

File: tendrl/ceph_integration/objects/event/__init__.py (hunk 18 above)
Comment (cybernth) [reply tendrl/ceph_integration/objects/event/__init__.py]: done

File: tendrl/ceph_integration/objects/pool/__init__.py (hunk 19 above)
Comment (cybernth) [reply tendrl/ceph_integration/objects/pool/__init__.py]: done

File: tendrl/ceph_integration/objects/pool/__init__.py (hunk 20 above)
Comment (cybernth) [reply tendrl/ceph_integration/objects/pool/__init__.py]: done

File: tendrl/ceph_integration/objects/pool/atoms/create.py (hunk 21 above)
Comment (cybernth) [reply tendrl/ceph_integration/objects/pool/atoms/create.py]: done

File: tendrl/ceph_integration/objects/pool/atoms/create.py (hunk 22 above)
Comment (cybernth) [reply tendrl/ceph_integration/objects/pool/atoms/create.py:11]: done

File: tendrl/ceph_integration/objects/pool/atoms/delete.py (hunk 23 above)
Comment (cybernth) [reply tendrl/ceph_integration/objects/pool/atoms/delete.py:3]: done

File: tendrl/ceph_integration/objects/pool/flows/delete_pool.py (hunk 24 above)
Comment (cybernth) [reply tendrl/ceph_integration/objects/pool/flows/delete_pool.py]: done

File: tendrl/ceph_integration/objects/sync_object/__init__.py (hunk 25 above)
Comment (cybernth) [reply tendrl/ceph_integration/objects/sync_object/__init__.py]: done

File: tendrl/ceph_integration/objects/tendrl_context/__init__.py (hunk 26 above)
Comment (cybernth) [reply tendrl/ceph_integration/objects/tendrl_context/__init__.py:3]: deferred as per the discussion

File: tendrl/ceph_integration/objects/tendrl_context/__init__.py (hunk 27 above)
Comment (cybernth) [reply tendrl/ceph_integration/objects/tendrl_context/__init__.py:22]: same as above

**Pull request merged**